src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -445,49 +445,33 @@ private static StringConcatException constantMismatch(Object[] constants,
      * most notably, the private String constructor that accepts byte[] arrays
      * without copying.
      */
-    private static MethodHandle generateMHInlineCopy(MethodType mt, List<String> elements) {
-
-        // Fast-path unary concatenations
-        if (elements.size() == 1) {
-            String s0 = elements.get(0);
-            if (s0 == null) {
-                return unaryConcat(mt.parameterType(0));
-            } else {
-                return MethodHandles.insertArguments(unaryConcat(Object.class), 0, s0);
-            }
-        }
-        // Fast-path binary concatenations
-        if (elements.size() == 2) {
-            // Two arguments
-            String s0 = elements.get(0);
-            String s1 = elements.get(1);
-
-            if (mt.parameterCount() == 2 &&
-                    !mt.parameterType(0).isPrimitive() &&
-                    !mt.parameterType(1).isPrimitive() &&
-                    s0 == null &&
-                    s1 == null) {
-                return simpleConcat();
-            } else if (mt.parameterCount() == 1) {
-                // One argument, one constant
-                String constant;
-                int constIdx;
-                if (s1 == null) {
-                    constant = s0;
-                    constIdx = 0;
-                } else {
-                    constant = s1;
-                    constIdx = 1;
-                }
-                if (constant.isEmpty()) {
+    private static MethodHandle generateMHInlineCopy(MethodType mt, List<String> constants) {
+        int paramCount = mt.parameterCount();
+        String suffix = constants.get(paramCount);
+
+        // Fast-path trivial concatenations
+        if (paramCount == 0) {
+            return MethodHandles.insertArguments(newStringifier(), 0, suffix == null ? "" : suffix);
+        }
+        if (paramCount == 1) {
+            String prefix = constants.get(0);
+            if (prefix == null) {
+                if (suffix == null) {
```

Comment:
Is it possible for prefix or suffix an empty string (non-null)?   The original code checks `if (constant.isEmpty())`

Reply:
No: Empty string constants would lead to adding a null to the list of constants in the prologue.